### PR TITLE
MG-2173 - Generate mocks with mockery for Twins service

### LINKED ISSUE #2173: Feature: Generate mocks with mockery for Twins service
### Is your feature request related to a problem? Please describe.

Requested feature is not related to a problem.

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Generate mocks with mockery and rewrite test for Twins service

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nyagamunene**
# What type of PR is this?
This is a feature because its add the following functionality: generates mocks for Twins.

## What does this do?
It generates the mocks for Twins service.

## Which issue(s) does this PR fix/relate to?

- Related Issue #2173 
- Resolves #2173 

## Have you included tests for your changes?
Yes, I have included tests

## Did you document any new/modified feature?
No, I have not updated the documentation

### CODE REVIEW

File: twins/api/http/endpoint_states_test.go [hunk 1]
```diff
@@ -40,27 +48,52 @@ type statesPageRes struct {
...
 
 	twin := twins.Twin{
 		Owner: email,
 	}
-	def := mocks.CreateDefinition(channels[0:2], subtopics[0:2])
+	def := CreateDefinition(channels[0:2], subtopics[0:2])
 	repoCall := auth.On("Identify", mock.Anything, &magistrala.IdentityReq{Token: token}).Return(&magistrala.IdentityRes{Id: testsutil.GenerateUUID(t)}, nil)
+	repoCall1 := twinRepo.On("Save", context.Background(), mock.Anything).Return(retained, nil)
+	repoCall2 := twinCache.On("Save", context.Background(), mock.Anything).Return(nil)
 	tw, err := svc.AddTwin(context.Background(), token, twin, def)
```
Comment (rodneyosodo) [inline twins/api/http/endpoint_states_test.go]: We don't need to mock this

File: twins/api/http/endpoint_states_test.go [hunk 2]
```diff
@@ -40,27 +48,52 @@ type statesPageRes struct {
...
-	message, err := mocks.CreateMessage(attr, recs)
+	CreateSenML(recs)
+	message, err := CreateMessage(attr, recs)
 	require.Nil(t, err, fmt.Sprintf("unexpected error: %s", err))
+	var testString []string
+	repoCall = auth.On("Identify", mock.Anything, &magistrala.IdentityReq{Token: token}).Return(&magistrala.IdentityRes{Id: testsutil.GenerateUUID(t)}, nil)
+	repoCall1 = twinCache.On("IDs", context.Background(), mock.Anything, mock.Anything).Return(testString, nil)
+	repoCall2 = twinRepo.On("RetrieveByAttribute", context.Background(), mock.Anything, mock.Anything).Return(testString, nil)
+	repoCall3 := twinRepo.On("SaveIDs", context.Background(), mock.Anything, mock.Anything, mock.Anything).Return(nil)
 	err = svc.SaveStates(context.Background(), message)
```
Comment (rodneyosodo) [inline twins/api/http/endpoint_states_test.go]: same

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Review by dborovcanin — DISMISSED** [review]

File: twins/api/http/endpoint_states_test.go [hunk 3]
```diff
@@ -40,152 +47,196 @@ type statesPageRes struct {
...
-	repoCall.Unset()
 	attr := def.Attributes[0]
-
 	recs := make([]senml.Record, numRecs)
-	mocks.CreateSenML(recs)
-	message, err := mocks.CreateMessage(attr, recs)
-	require.Nil(t, err, fmt.Sprintf("unexpected error: %s", err))
-	err = svc.SaveStates(context.Background(), message)
+	CreateSenML(recs)
+	_, err := CreateMessage(attr, recs)
```
Comment (rodneyosodo) [inline twins/api/http/endpoint_states_test.go]: If we are not using the output from `CreateMessage` why are we using it?

File: twins/api/http/endpoint_states_test.go [hunk 4]
```diff
@@ -40,152 +47,196 @@ type statesPageRes struct {
...
+			page: twins.StatesPage{
+				States: convStat(data[0:5]),
+			},
+			err: nil,
 		},
 	}
 
 	for _, tc := range cases {
 		repoCall := auth.On("Identify", mock.Anything, mock.Anything).Return(&magistrala.IdentityRes{Id: testsutil.GenerateUUID(t)}, nil)
+		repoCall1 := stateRepo.On("RetrieveAll", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(tc.page, tc.err)
```
Comment (rodneyosodo) [inline twins/api/http/endpoint_states_test.go]: Avoid the use of `mock.Anything`

File: twins/api/http/endpoint_states_test.go [hunk 5]
```diff
@@ -40,152 +47,196 @@ type statesPageRes struct {
...
 			res:    data[0:5],
+			page: twins.StatesPage{
+				States: convStat(data[0:5]),
+			},
+			err: nil,
 		},
 	}
 
 	for _, tc := range cases {
 		repoCall := auth.On("Identify", mock.Anything, mock.Anything).Return(&magistrala.IdentityRes{Id: testsutil.GenerateUUID(t)}, nil)
```
Comment (rodneyosodo) [inline twins/api/http/endpoint_states_test.go]: Add unhappy flows to simulate authentication failure

File: twins/api/http/endpoint_twins_test.go [hunk 6]
```diff
@@ -184,11 +191,14 @@ func TestAddTwin(t *testing.T) {
 			auth:        token,
 			status:      http.StatusBadRequest,
 			location:    "",
+			err:         svcerr.ErrMalformedEntity,
 		},
 	}
 
 	for _, tc := range cases {
 		repoCall := auth.On("Identify", mock.Anything, &magistrala.IdentityReq{Token: tc.auth}).Return(&magistrala.IdentityRes{Id: testsutil.GenerateUUID(t)}, nil)
+		repoCall1 := twinRepo.On("Save", mock.Anything, mock.Anything).Return(retained, tc.err)
```
Comment (rodneyosodo) [inline twins/api/http/endpoint_twins_test.go]: Try and reduce the usage of `mock.Anything`. This applies to other PRs and also other cases in this PR

File: twins/api/http/endpoint_twins_test.go [hunk 7]
```diff
@@ -259,58 +271,68 @@ func TestUpdateTwin(t *testing.T) {
...
 			contentType: contentType,
 			auth:        token,
 			status:      http.StatusMethodNotAllowed,
+			err:         svcerr.ErrMalformedEntity,
 		},
 	}
 
 	for _, tc := range cases {
 		repoCall := auth.On("Identify", mock.Anything, &magistrala.IdentityReq{Token: tc.auth}).Return(&magistrala.IdentityRes{Id: testsutil.GenerateUUID(t)}, nil)
+		repoCall1 := twinRepo.On("RetrieveByID", mock.Anything, tc.id).Return(twins.Twin{}, tc.err)
```
Comment (rodneyosodo) [inline twins/api/http/endpoint_twins_test.go]: Simulate different failures have error variables for `retrieveByIDErr`, `updateErr` and `cacheErr`

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: twins/mongodb/twins_test.go [hunk 8]
```diff
@@ -386,3 +387,26 @@ func TestTwinsRemove(t *testing.T) {
 		assert.Equal(t, tc.err, err, fmt.Sprintf("%s: expected %s got %s\n", tc.desc, tc.err, err))
 	}
 }
+
+// CreateDefinition creates twin definition.
+func CreateDefinition(channels, subtopics []string) twins.Definition {
```
Comment (rodneyosodo) [inline twins/mongodb/twins_test.go]: We can define these 2 functions in mocks the reuse them in mocks service tests

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Review by rodneyosodo — DISMISSED** [review]

**Comment by WashingtonKK** [comment]
Can we upgrade to the latest version of mockery (v2.42.3)?

**Review by rodneyosodo — DISMISSED** [review]

File: twins/api/http/endpoint_states_test.go [hunk 9]
```diff
@@ -215,3 +291,16 @@ func createStateResponse(id int, tw twins.Twin, rec senml.Record) stateRes {
 		Payload:    map[string]interface{}{rec.BaseName: nil},
 	}
 }
+
+func convStat(data []stateRes) []twins.State {
```
Comment (rodneyosodo) [inline twins/api/http/endpoint_states_test.go]:
```suggestion
func convState(data []stateRes) []twins.State {
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Review by rodneyosodo — APPROVED** [review]

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**